src/uct/ib/ud/base/ud_iface.c

```diff
@@ -313,25 +313,82 @@ uct_ud_iface_create_qp(uct_ud_iface_t *self, const uct_ud_iface_config_t *config
     return UCS_ERR_INVALID_PARAM;
 }
 
+static inline void uct_ud_iface_async_progress(uct_ud_iface_t *iface)
+{
+    uct_ud_iface_ops_t *ops =
+        ucs_derived_of(iface->super.ops, uct_ud_iface_ops_t);
+    unsigned ev_count;
+
+    ev_count = ops->async_progress(iface);
+    if (ev_count > 0) {
+        uct_ud_iface_raise_pending_async_ev(iface);
+    }
+}
+
+static void uct_ud_iface_async_handler(int fd, int events, void *arg)
+{
+    uct_ud_iface_t *iface = arg;
+
+    uct_ud_enter(iface);
```

Comment:
no need to take lock

Reply:
I think we can't avoid taking lock here - same for `uct_ud_iface_timer()`

Reply:
seems it's called from async handler/timer, so it's already under lock